lib/base/io-engine.cpp

```diff
@@ -20,56 +20,41 @@ CpuBoundWork::CpuBoundWork(boost::asio::yield_context yc)
 	: m_Done(false)
 {
 	auto& ioEngine (IoEngine::Get());
+	auto& sem (ioEngine.m_CpuBoundSemaphore);
+	std::unique_lock<std::mutex> lock (sem.Mutex);
 
-	for (;;) {
-		auto availableSlots (ioEngine.m_CpuBoundSemaphore.fetch_sub(1));
+	if (sem.FreeSlots) {
+		--sem.FreeSlots;
+		return;
+	}
 
-		if (availableSlots < 1) {
-			ioEngine.m_CpuBoundSemaphore.fetch_add(1);
-			IoEngine::YieldCurrentCoroutine(yc);
-			continue;
-		}
+	auto cv (Shared<AsioConditionVariable>::Make(ioEngine.GetIoContext()));
 
-		break;
-	}
-}
+	sem.Waiting.emplace(cv);
+	lock.unlock();
 
-CpuBoundWork::~CpuBoundWork()
-{
-	if (!m_Done) {
-		IoEngine::Get().m_CpuBoundSemaphore.fetch_add(1);
+	try {
+		cv->Wait(yc);
+	} catch (...) {
+		Done();
```

Comment:
Why is `Done()` called here? Wouldn't this release a slot that was never acquired?

Reply:
One of the pillars of the whole logic:

A regular CpuBoundWork#Done() from CpuBoundWork#~CpuBoundWork() calls AsioConditionVariable#Set() and expects it to successfully finish AsioConditionVariable#Wait() and CpuBoundWork#CpuBoundWork(). The latter implies a later CpuBoundWork#~CpuBoundWork() which again calls CpuBoundWork#Done(). But if AsioConditionVariable#Wait() throws, I can call CpuBoundWork#Done() now or never.

Reply:
But still, this means that more coroutines can simultaneously acquire `CpuBoundMutex` than what is permitted by `ioEngine.m_CpuBoundSemaphore`.